### ISSUE TITLE
DNS Aliases

### ISSUE THREAD

**Issue by AndrewDryga** [CloudFire-LLC/cloudfire#2801]
We can implement an internal DNS scheme which will give the client device much more control about how it wants to route the traffic, the generic scheme can look something like this: `{ip}?.{alias|"i"}?.{site_id|"g"}.fz.local`

There are many use cases for it:

1. Accessing a single IP address in a specific site: `10-10-10-10.i.mysite.fz.local` - access `10.10.10.10` at site `mysite`.
2. Accessing a single IP across all sites that define a resource with such address: `10-10-10-10.i.g.fz.local`
3. Access a single IP in a named CIDR resource `myapp` at site `mysite`: `10-10-10-10.myapp.mysite.fz.local`
4. Access one of IPs (load balanced by DNS resolver) in a named CIDR resource `metabase` at site `mysite`: `metabase.mysite.fz.local`
5. Same as (4) but DNS-balanced across multiple sites: `myapp.g.fz.local`

This will solve edge cases when we have multiple VPCs available by the same user, and allow the user to decide if overlapping VPCs should be treated as HA-zones or if it wants to access a specific site.

Alternative is name labeling:
```
i-10-10-10-10.
s-mysite.
r-myapp.

i-10-10-10-10.r-myapp.s-mysite.fz.local
r-myapp.s-mysite.fz.local
s-mysite.fz.local
i-10-10-10-10.r-myapp.fz.local
i-10-10-10-10.fz.local
r-myapp.fz.local
```

**Comment by AndrewDryga** [comment]
We can potentially remove `i.` and `g.` that I added to remove more chances of collisions and reserved names too and make DNS addresses even shorter, this is especially easy if we force everything to be aliased and won't take `ip` part. This will bring down the scheme to something as simple as:

```
myresource.mysite.fz.local
myresource.fz.local
```

**Comment by AndrewDryga** [comment]
Crazy idea for future, allow exposing resources via our POPs by using public tld `myresource.mysite.fz.net`

**Comment by jamilbk** [comment]
Interesting idea. I suspect we'll still need to support raw CIDR / IP addressing for some use cases, but this would be handy for dealing with IP range conflicts if they arise.

Will add to Public Roadmap board for more visibility.

We could save this for a post-beta or post-1.0 feature launch "Aliases".

**Comment by Yasamato** [comment]
My understanding of this RFC seems to be lacking, but I see a potential conflict with my knowledge of how DNS works.

Scenario:
You have a service on `mysite` with the local IP `10.10.10.10`. In the local network is a gateway deployed as well. Now you enter into the browser `http://10-10-10-10.mysite.fz.local`.

The browser send now the a `HTTP GET` to the provided IP. As it's a HTTP request, you could extract the requested domain from the request and correctly route it, but what if it's an arbitrary other protocol where one cannot gain insights from?

Assume that e.g. `10-10-10-10.mysite.fz.local` resolves to some arbitrary IP, would this be `10.10.10.10` or the IP of the gateway at `mysite`? If it's `10.10.10.10`, what if in the same local network there exists the same IP? How would the client then determine that it should route to `10.10.10.10` in `mysite` instead of the local one?

One workaround would be to wrap it into an ipv6 and encode it there, but then we run into the same issue when we have an ipv6 network? Would the schema be `fe80-0-0-0-0-10.mysite.fz.local`?

This would require unique subnets for each site to avoid the concerns I outlined.

**Comment by jamilbk** [comment]
@Yasamato Good catch, and great question. We have a todo item to write a blog post explaining how DNS works in Firezone 1.0, but the short answer is: we run a tiny proxy in each client that generates dummy addresses on-the-fly at the time the DNS query matches.

These dummy addresses are generated from the `100.64.96.0/11` range (IPv4) and `fd00:2021:1111:8000/107` range (IPv6) and won't conflict with other resources in your Firezone Account.

Taking your example:

* `10-10-10-10.mysite.fz.local` would resolve to `100.64.96.0.1` in the client
* `10-10-10-10.othersite.fz.local` would resolve to `100.64.96.0.2` in the client

Then, when the client sees a packet to one of these `100.64.x.x` IPs, it will know which site (and pool of gateways) to request an authorization (and holepunch, connection, etc) for.

With this DNS alias approach we can route packets to as many conflicting private subnets as the administrator wishes to define.

See https://github.com/firezone/firezone/issues/2661 for more discussion on this design.

**Comment by Yasamato** [comment]
Thanks for explaining @jamilbk! With the approach of a reserved subnet for temporal ip-mapping I do not see any issues at the moment besides dns caching on the client side. Such a reserved subnet needs to be communicated thou and the network design on the sites have to comply with the network requirements of firezone.

**Comment by thomaseizinger** [comment]
We may want to consider putting all these domains under `firezone.dev` (or some other domain we control) to make sure TLS works if the user wants to use TLS.

**Comment by thomaseizinger** [comment]
> We may want to consider putting all these domains under `firezone.dev` (or some other domain we control) to make sure TLS works if the user wants to use TLS.

Actually, this is harder than I thought it would be. We'd need to be a CA and issue certificates for those domains.

**Comment by thomaseizinger** [comment]
`.local` is for mDNS and will likely cause problems. `.internal` has just been reserved though so that could be an option: https://www.icann.org/en/board-activities-and-meetings/materials/approved-resolutions-special-meeting-of-the-icann-board-29-07-2024-en#section2.a
